fortanix-vme/aws-nitro-enclaves/aws-nitro-blobs-downloader/src/lib.rs

```diff
@@ -0,0 +1,116 @@
+const NSM_X86_64_URL: &str = "https://raw.githubusercontent.com/aws/aws-nitro-enclaves-cli/ec130adc1fd86f2489f482d3f4a02676d3a748a7/blobs/x86_64/nsm.ko";
+const NSM_X86_64_SHA256: &str = "48904e520db2541ca4378da29d85791749408febc81987ade56cc5c556bd90df";
+
+struct Blob {
+    url: &'static str,
+    sha256: &'static str,
+}
+
+const BLOBS: [Blob; 5] = [
+    Blob {
+        url: BZIMAGE_X86_64_URL,
+        sha256: BZIMAGE_X86_64_SHA256,
+    },
+    Blob {
+        url: BZIMAGE_CONFIG_X86_64_URL,
+        sha256: BZIMAGE_CONFIG_X86_64_SHA256,
+    },
+    Blob {
+        url: CMDLINE_X86_64_URL,
+        sha256: CMDLINE_X86_64_SHA256,
+    },
+    Blob {
+        url: INIT_X86_64_URL,
+        sha256: INIT_X86_64_SHA256,
+    },
+    Blob {
+        url: NSM_X86_64_URL,
+        sha256: NSM_X86_64_SHA256,
+    },
+];
+
+impl Blob {
+    fn download(&self, client: &Client, output_dir: &Path) -> anyhow::Result<PathBuf> {
+        let filename = self
+            .url
+            .rsplit('/')
+            .next()
+            .context("Failed to get filename")?;
+        let fullpath = output_dir.join(filename);
+        if fullpath.exists() {
+            let file_hash = Self::hash_file(&fullpath)?;
+            match self.verify_hash(&file_hash) {
+                Ok(_) => {
+                    debug!("Hash of file {:?} verified.", fullpath);
+                    return Ok(fullpath);
+                }
+                Err(_) => warn!("Hash check failed on {:?}, re-downloading ...", fullpath),
+            }
+        }
+
+        info!("Downloading '{}' to {:?}", self.url, fullpath);
+        let mut response = client.get(self.url).send()?;
+        let mut fd = File::create(fullpath.clone())?;
+
+        response.copy_to(&mut fd)?;
+        // Close the file before calculating hash of it.
+        drop(fd);
+
+        let hash = Self::hash_file(&fullpath)?;
+        self.verify_hash(&hash)
+            .context("Expected hash mismatched")?;
+        Ok(fullpath)
+    }
+
+    fn hash_file(path: &PathBuf) -> anyhow::Result<[u8; 32]> {
+        let mut file = File::open(path)?;
+        let mut hasher = Sha256::new();
+        io::copy(&mut file, &mut hasher)?;
+        let hash = hasher.finalize();
+        Ok(hash.into())
+    }
+
+    fn verify_hash(&self, hash: &[u8]) -> anyhow::Result<()> {
+        let expected = hex::decode(self.sha256)?;
+        if expected != hash {
+            anyhow::bail!("Hash check failed");
+        }
+
+        Ok(())
+    }
+}
+
+pub fn download_blobs(output_dir: &Path) -> anyhow::Result<Vec<String>> {
+    if !output_dir.is_dir() {
+        anyhow::bail!("{:?} must be a directory.", output_dir);
+    }
+
+    let client = reqwest::blocking::Client::new();
+    let mut blobs = Vec::new();
+    for blob in &BLOBS {
+        let blob_file = blob.download(&client, output_dir)?;
+        blobs.push(blob_file.to_string_lossy().to_string());
```

Comment:
Can you return a `Vec` of `PathBuf` instead? In most cases if you want to do anything with the path, you need to turn it into a `Path` again, and turning it into a string is easy anyway.